### PR TITLE
[mkt] separated store from matcher

### PR DESCRIPTION
**Pull request by tworec**
Offer and Demand store as a separate structure to handle persistence. Matcher is now handling only Offers (subscriptions, subscriptions, and other related stuff).

### CODE REVIEW

File: core/market/decentralized/src/matcher/store.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+    pub async fn create_offer(
+        &self,
+        id: &Identity,
+        offer: &ClientOffer,
+    ) -> Result<Offer, OfferError> {
+        let creation_ts = Utc::now().naive_utc();
+        // TODO: provider agent should set expiration.
+        let expiration_ts = creation_ts + Duration::hours(24);
+        let offer = Offer::from_new(offer, &id, creation_ts, expiration_ts);
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:31]: I wanted to enclose all time related stuff within single class.

File: core/market/decentralized/src/market.rs [hunk 2]
```diff
@@ -55,7 +57,7 @@ impl MarketService {
             RequestorNegotiationEngine::new(db.clone(), listeners.proposal_receiver)?;
 
         Ok(MarketService {
-            matcher: Arc::new(matcher),
+            matcher,
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/market.rs:60]: we do not need clone at all. removing it

File: core/market/decentralized/src/matcher/store.rs [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                // should be removed.
+                // We are sure that we don't remove our Offer here, because we would got
+                // `AlreadyUnsubscribed` error from `checked_unsubscribe_offer`,
+                // as it was already invoked before broadcast in `unsubscribe_offer`.
+                // TODO: Maybe we should add check here, to be sure, that we don't remove own Offers.
+                log::debug!("Removing unsubscribed Offer [{}].", id);
+                let _ = self.remove_offer(id).await.map_err(|e| {
+                    log::warn!("Failed to remove offer [{}] during unsubscribe: {}", id, e);
+                });
+                Ok(Propagate::Yes)
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:90]: Few hours after pushing this change I've reconsidered it and went to the same conclusion.

File: core/market/decentralized/src/matcher/store.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/store.rs:31]: These 2 timestamps have different semantic.
It is additional value, that you can set expiration. But it is decrease in value, that you can set creation timestamp.
But it's minor.

[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:31]: I imagine that you can use some additional logic in setting creation, expiration and validation times eg. using some arbitrary precision. So I'd like to have it all in single place.

[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:31]: Furthermore having creation_ts parametrized you can use it to set relative expiration_ts

File: core/market/decentralized/src/matcher/store.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+            Ok(true) => Ok(()),
+            Ok(false) => Err(OfferError::UnexpectedError(format!(
+                "Offer [{}] marked as unsubscribed, but not removed",
+                id
+            ))),
+            Err(e) => Err(OfferError::RemoveError(e, id.clone()).into()),
+        }
+    }
+
+    pub async fn checked_store_offer(&self, offer: Offer) -> Result<bool, OfferError> {
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:92]: `save_offer` is private. It is used internally in a few places.

as to the naming -- I agree. I couldn't find anything good. You are the first reader and from your opinion I assume it enough to remove `checked` as you suggest.

File: core/market/decentralized/src/matcher/store.rs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+            Ok(OfferState::Expired(_)) => Err(OfferError::Expired(id.clone())),
+            Ok(OfferState::NotFound) => Err(OfferError::NotFound(id.clone())),
+        }
+    }
+
+    /// We store only our Offers to keep history. Offers from other nodes
+    /// should be removed.
+    /// This is meant to be called upon receiving unsubscribe broadcast. To work correctly
+    /// it assumes `mark_offer_unsubscribed` was invoked before broadcast in `unsubscribe_offer`.
+    pub async fn checked_remove_offer(&self, id: &SubscriptionId) -> Result<(), OfferError> {
```
[the remark this answers is not in the document]
Comment (tworec) [reply core/market/decentralized/src/matcher/store.rs:76]: just with the naming. There is no internal helper function. Will remove `checked` then

File: core/market/decentralized/src/matcher/store.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply core/market/decentralized/src/matcher/store.rs:92]: Yes. It would be enough for me to accept this PR ;)

**Pull request merged**